pkg/reboot/calculator.go

```diff
@@ -23,102 +23,100 @@ const (
-	s.timeToAssumeNodeRebooted = timeToAssumeNodeRebooted
+	specRebootDurationSeconds := config.Spec.SafeTimeToAssumeNodeRebootedSeconds
+	if specRebootDurationSeconds != nil {
+		specRebootDuration := time.Duration(*specRebootDurationSeconds) * time.Second
+		// In case users specified a lower reboot time, ignore it
+		if specRebootDuration < minimumCalculatedRebootDuration {
+			log.V(0).Info("Warning: Ignoring specified SafeTimeToAssumeNodeRebootedSeconds because it's lower than the calculated minimum safe reboot time", "specified time in seconds", specRebootDuration.Seconds(), "calculated minimum time in seconds", minimumCalculatedRebootDuration.Seconds())
+			// TODO event
+			return minimumCalculatedRebootDuration, nil
+		}
+		log.Info("Using specified SafeTimeToAssumeNodeRebootedSeconds because it's greater than the calculated minimum safe reboot time", "specified time in seconds", specRebootDuration.Seconds(), "calculated minimum time in seconds", minimumCalculatedRebootDuration.Seconds())
+		return specRebootDuration, nil
+	}
+	log.Info("No SafeTimeToAssumeNodeRebootedSeconds specified, using calculated minimum safe reboot time", "calculated minimum time in seconds", minimumCalculatedRebootDuration)
+	return minimumCalculatedRebootDuration, nil
 }
 
-func (s *safeTimeCalculator) Start(_ context.Context) error {
-	return s.calcMinTimeAssumeRebooted()
-}
+func (r *rebootDurationCalculator) calculateMinimumRebootDuration(k8sClient client.Client, ctx context.Context, cfg *v1alpha1.SelfNodeRemediationConfig, watchdogTimeout time.Duration) (time.Duration, error) {
+
+	spec := cfg.Spec
+
+	// The reboot duration needs be at least the time we know we need for determining a node issue and trigger the reboot!
 
-func (s *safeTimeCalculator) calcMinTimeAssumeRebooted() error {
-	if !s.isAgent {
-		return nil
-	}
-	// The reboot time needs be at least the time we know we need for determining a node issue and trigger the reboot!
 	// 1. time for determine node issue
-	minTime := (s.apiCheckInterval+s.apiServerTimeout)*time.Duration(s.maxErrorThreshold) + MaxTimeForNoPeersResponse
-	// 2. time for asking peers (10% batches + 1st smaller batch)
-	minTime += time.Duration(s.calcNumOfBatches()) * (s.peerDialTimeout + s.peerRequestTimeout)
-	// 3. watchdog timeout
-	if s.wd != nil {
-		minTime += s.wd.GetTimeout()
+	// a) API check duration ...
+	minTime := spec.ApiCheckInterval.Duration + spec.ApiServerTimeout.Duration
+	// b) ... times error threshold ...
+	minTime *= time.Duration(spec.MaxApiErrorThreshold)
+	// c) ... plus peer timeout
+	minTime += MaxTimeForNoPeersResponse
+
+	// 2. plus time for asking peers (10% batches + 1st smaller batch)
+	numBatches, err := r.calcNumOfBatches(k8sClient, ctx)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to calculate number of batches")
 	}
-	// 4. some buffer
-	minTime += 15 * time.Second
-	s.log.Info("calculated minTimeToAssumeNodeRebooted is:", "minTimeToAssumeNodeRebooted", minTime)
```

Comment:
keeping the log with the calculated time might be useful for debugging issues

Reply:
it's logged at the caller in every case